gems/smithy/lib/smithy/model/flattener.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Model
+    # @api private
+    class Flattener
+      def initialize(model)
+        @model = model
+      end
+
+      def shape(id)
+        shape = @model['shapes'][id]
+        return shape unless shape['mixins']
+
+        shape['mixins'].reverse_each do |mixin|
+          mixin_shape = shape(mixin['target'])
+          shape = deep_merge(mixin_shape, shape, exclude_traits(mixin_shape))
+          apply_member_traits(id, shape)
+          shape.delete('mixins')
+        end
+
+        shape
+      end
+
+      private
+
+      def exclude_traits(shape)
+        [
+          'smithy.api#mixin',
+          *shape.fetch('traits', {}).fetch('smithy.api#mixin', {}).fetch('localTraits', [])
+        ]
+      end
+
+      def deep_merge(hash1, hash2, exclude_traits = [], context = nil)
+        hash1 = hash1.dup
+        if hash1['traits']
+          hash1['traits'] = hash1['traits'].except(*exclude_traits)
+          hash1.delete('traits') if hash1['traits'].empty?
+        end
+        deep_merge!(hash1, hash2, exclude_traits, context)
+      end
+
+      def deep_merge!(hash1, hash2, exclude_traits, context)
```

Comment:
The difference between deep_merge and deep_merge! is confusing - maybe these could use comments?

Reply:
One returns a new hash and the other modifies the current hash, just like merge and merge!. I got inspiration from the rails implementations. I can add that here.